### PR TITLE
documentation: (toy) add optimisation steps to Toy

### PR DESCRIPTION
**Pull request by superlopuh**
This is technically in Toy but we're currently using this pipeline for the riscv paper experiments, so would benefit from this change for critical work.

### CODE REVIEW

File: docs/Toy/toy/compiler.py [hunk 1]
```diff
@@ -124,20 +124,32 @@ def transform(
...
     ReconcileUnrealizedCastsPass().apply(ctx, module_op)
 
-    DeadCodeElimination().apply(ctx, module_op)
-
     module_op.verify()
 
     if target == "riscv":
         return
 
+    CanonicalizePass().apply(ctx, module_op)
```
Comment (superlopuh) [inline docs/Toy/toy/compiler.py:132]: optimises things that don't need register allocation, like constant folding etc

File: docs/Toy/toy/compiler.py [hunk 2]
```diff
@@ -124,20 +124,32 @@ def transform(
...
+        return
+
     RISCVRegisterAllocation().apply(ctx, module_op)
 
     module_op.verify()
 
     if target == "riscv-regalloc":
         return
 
+    CanonicalizePass().apply(ctx, module_op)
```
Comment (superlopuh) [inline docs/Toy/toy/compiler.py:146]: optimises things that need register allocation, like redundant moves etc

Comment (nazavode) [reply docs/Toy/toy/compiler.py:146]:
```suggestion
    # Perform optimizations that depend on register allocation (e.g.: redundant moves):
    CanonicalizePass().apply(ctx, module_op)
```

File: docs/Toy/toy/compiler.py (hunk 1 above)
Comment (nazavode) [reply docs/Toy/toy/compiler.py:132]:
```suggestion
    # Perform optimizations that don't depend on register allocation (e.g.: constant folding):
    CanonicalizePass().apply(ctx, module_op)
```

**Review by nazavode — APPROVED** [review]
Nice, left a couple of suggested changes that bring your own comments in source for posterity.

**Pull request merged**